odpf/guardian/guardian.proto

```diff
@@ -0,0 +1,375 @@
+        option (google.api.http) = {
+            put: "/appeals/{id}/revoke"
+        };
+    }
+
+    rpc CreateAppeal(CreateAppealRequest) returns (CreateAppealResponse) {
+        option (google.api.http) = {
+            post: "/appeals"
+            body: "*"
+        };
+    }
+
+    rpc ListApprovals(ListApprovalsRequest) returns (ListApprovalsResponse) {
+        option (google.api.http) = {
+            get: "/appeals/approvals"
+        };
+    }
+
+    rpc UpdateApproval(UpdateApprovalRequest) returns (UpdateApprovalResponse) {
+        option (google.api.http) = {
+            post: "/appeals/{id}/approvals/{approval_name}"
+            body: "*"
+        };
+    }
+}
+
+message ListProvidersRequest {}
+
+message ListProvidersResponse {
+    repeated Provider providers = 1;
+}
+
+message CreateProviderRequest {
+    Provider provider = 1;
+}
+
+message CreateProviderResponse {
+    Provider provider = 1;
+}
+
+message UpdateProviderRequest {
+    uint32 id = 1;
+    Provider provider = 2;
+}
+
+message UpdateProviderResponse {
+    Provider provider = 1;
+}
+
+message ListPoliciesRequest {}
+
+message ListPoliciesResponse {
+    repeated Policy policies = 1;
+}
+
+message CreatePolicyRequest {
+    Policy policy = 1;
+}
+
+message CreatePolicyResponse {
+    Policy policy = 1;
+}
+
+message UpdatePolicyRequest {
+    string id = 1;
+    Policy policy = 2;
+}
+
+message UpdatePolicyResponse {
+    Policy policy = 1;
+}
+
+message ListResourcesRequest {}
+
+message ListResourcesResponse {
+    repeated Resource resources = 1;
+}
+
+message UpdateResourceRequest {
+    uint32 id = 1;
+    Resource resource = 2;
+}
+
+message ListAppealsRequest {
+    string user = 1;
+}
+
+message ListAppealsResponse {
+    repeated Appeal appeals = 1;
+}
+
+message GetAppealRequest {
+    uint32 id = 1;
+}
+
+message GetAppealResponse {
+    Appeal appeal = 1;
+}
+
+message CancelAppealRequest {
+    uint32 id = 1;
+}
+
+message CancelAppealResponse {
+    Appeal appeal = 1;
+}
+
+message RevokeAppealRequest {
+    uint32 id = 1;
+}
+
+message RevokeAppealResponse {
+    Appeal appeal = 1;
+}
+
+message CreateAppealRequest {
+    string user = 1;
+
+    message Resource {
+        uint32 id = 1;
+        string role = 2;
+        google.protobuf.Struct options = 3;
+    }
+    repeated Resource resources = 2;
+}
+
+message CreateAppealResponse {
+    repeated Appeal appeals = 1;
+}
+
+message ListApprovalsRequest {
+    string user = 1;
+}
+
+message ListApprovalsResponse {
+    repeated Approval approvals = 1;
+}
+
+message UpdateApprovalRequest {
+    uint32 id = 1;
+    string approval_name = 2;
+
+    message Action {
+        string action = 1;
+    }
+    Action action = 3;
+}
+
+message UpdateApprovalResponse {
+    Appeal appeal = 1;
+}
+
+// Provider contains information about external data provider such as BigQuery, Metabase, etc., credentials, policy, and allowed roles 
+message Provider {
+    uint32 id = 1;
```

Comment:
I see we are using int ids everywhere, are these DB auto-increment ids? Why not use UUID?

Reply:
Yes DB auto-increment, no specific reason why we're not using UUID, just following traditional ways for creating entity. What would make it better to use UUID?

Reply:
- can't enumerate values
- can't infer how many values are present in each table
- don't have to talk to the database to construct a value, useful in certain ORM situations
- perfectly indexable by most dbs
- easier to shard when we need to distribute data into multiple dbs
- unique across multiple tables/databases/space-time continuum
- costly on size but storage is cheap